src/main/scala/nodes/util/CommonSparseFeatures.scala

```diff
@@ -34,8 +40,25 @@ case class CommonSparseFeatures[T : ClassTag](numFeatures: Int) extends Estimato
     val featureFrequenciesWithUniqueId = featureOccurrences.reduceByKey {
       (x, y) => (x._1 + y._1, Math.min(x._2, y._2))
     }
-    val mostCommonFeatures = featureFrequenciesWithUniqueId.top(numFeatures)(ordering).map(_._1)
+    val mapRDDs = featureFrequenciesWithUniqueId mapPartitions { items =>
+      // Priority keeps the largest elements, so let's reverse the ordering.
+      Iterator.single(takeOrdered(items, numFeatures)(ordering.reverse))
+    }
+    val mostCommonFeatures = mapRDDs.treeReduce(merge).map(_._1)
+
     val featureSpace = mostCommonFeatures.zipWithIndex.toMap
     new SparseFeatureVectorizer(featureSpace)
   }
+
+  /**
+   * Returns the first K elements from the input as defined by the specified implicit Ordering[T]
+   * and maintains the ordering.
+   */
+  def takeOrdered[T](input: Iterator[T], num: Int)(implicit ord: Ordering[T]): Seq[T] = {
+    val ordering = new com.google.common.collect.Ordering[T] {
```

Comment:
I don't think we explicitly depend on guava in our build.sbt and we seem to be getting it through some transitive dependencies ?